src/backend/distributed/operations/auto_shard_split.c

```diff
@@ -0,0 +1,370 @@
+	int64 shardmaxvalue;
+	int64 shardid;
+	int32 nodeid;
+	char *tablename;
+	char *distributionColumn;
+	char *datatype;
+	char *shardname;
+	Oid tableId;
+	Oid distributionColumnId;
+}ShardInfoData;
+typedef ShardInfoData *ShardInfo;
+
+
+/*
+ * It throws an error if a concurrent automatic shard split or Rebalance operation is happening.
+ */
+void
+ErrorOnConcurrentOperation()
+{
+	int64 jobId = 0;
+	if (HasNonTerminalJobOfType("rebalance", &jobId))
+	{
+		ereport(ERROR, (
+					errmsg("A rebalance is already running as job %ld", jobId),
+					errdetail("A rebalance was already scheduled as background job"),
+					errhint("To monitor progress, run: SELECT * FROM "
+							"citus_rebalance_status();")));
+	}
+	if (HasNonTerminalJobOfType("Automatic Shard Split", &jobId))
+	{
+		ereport(ERROR, (
+					errmsg("An automatic shard split is already running as job %ld",
+						   jobId),
+					errdetail(
+						"An automatic shard split was already scheduled as background job")));
+	}
+}
+
+
+/*
+ * For a given SplitPoints , it creates the SQL query for the shard Splitting
+ */
+StringInfo
+GetShardSplitQuery(ShardInfo shardinfo, List *SplitPoints , char* shardSplitMode)
+{
+	StringInfo splitQuery = makeStringInfo();
+
+	int64 length = list_length(SplitPoints);
+	appendStringInfo(splitQuery, "SELECT citus_split_shard_by_split_points(%ld, ARRAY[",
+					 shardinfo->shardid);
+
+	for (int i = 0; i < length - 1; i++)
+	{
+		appendStringInfo(splitQuery, "'%d',", list_nth_int(SplitPoints, i));
+	}
+	appendStringInfo(splitQuery, "'%d'], ARRAY[", list_nth_int(SplitPoints,
+																		  length - 1));
+
+	for (int i = 0; i < length; i++)
+	{
+		appendStringInfo(splitQuery, "%d,", shardinfo->nodeid);
+	}
+	appendStringInfo(splitQuery, "%d], %s)", shardinfo->nodeid, quote_literal_cstr(shardSplitMode));
+
+	return splitQuery;
+}
+
+
+/*
+ * It creates a background job for citus_split_shard_by_split_points and executes it in background.
+ */
+void
+ExecuteSplitBackgroundJob(int64 jobid, ShardInfo shardinfo, List *SplitPoints , char* shardSplitMode)
+{
+	StringInfo splitQuery = makeStringInfo();
+	splitQuery = GetShardSplitQuery(shardinfo, SplitPoints , shardSplitMode);
+	ereport(LOG, (errmsg(splitQuery->data)));
+	int32 nodesInvolved[1];
+	nodesInvolved[0] = shardinfo->nodeid;
+	Oid superUserId = CitusExtensionOwner();
+	BackgroundTask *task = ScheduleBackgroundTask(jobid, superUserId, splitQuery->data, 0,
+												  NULL, 1, nodesInvolved);
+}
+
+
+/*
+ * It executes a query to find the average hash value in a shard considering rows with a limit of 10GB .
+ * If there exists a hash value it is returned otherwise shardminvalue-1 is returned.
+ */
+int64
+ExecuteAvgHashQuery(ShardInfo shardinfo)
+{
+	StringInfo AvgHashQuery = makeStringInfo();
+	appendStringInfo(AvgHashQuery, "SELECT avg(h)::int,count(*)"
+								   " FROM (SELECT worker_hash(%s) h FROM %s TABLESAMPLE SYSTEM(least(10, 100*10000000000/citus_total_relation_size(%s)))"
```

Comment:
this query is not wrong, though the call to citus_total_relation_size will get pushed down to each of the shards, and query all the worker nodes for their sizes. 

It would be good to get the size using DistributedTableSize (see the citus_total_relation_size implementation in metadata_utility.c), and pass the value to the query

Reply:
used DistributedTableSize for now , trying to get from the query as well